executor/aggfuncs/func_count_distinct.go

```
@@ -433,3 +435,387 @@ func appendJSON(encodedBytes, _ []byte, val json.BinaryJSON) []byte {
+func NewPartialResult4ApproxCountDistinct() *partialResult4ApproxCountDistinct {
+	p := &partialResult4ApproxCountDistinct{}
+	p.reset()
+	return p
+}
+
+func (p *partialResult4ApproxCountDistinct) InsertHash64(x uint64) {
+	// no need to rehash, just cast into uint32
+	p.insertHash(approxCountDistinctHashValue(x))
+}
+
+func (p *partialResult4ApproxCountDistinct) alloc(newSizeDegree uint8) {
+	p.size = 0
+	p.skipDegree = 0
+	p.hasZero = false
+	p.buf = make([]approxCountDistinctHashValue, uint32(1)<<newSizeDegree)
+	p.sizeDegree = newSizeDegree
+}
+
+func (p *partialResult4ApproxCountDistinct) reset() {
+	p.alloc(uniquesHashSetInitialSizeDegree)
+}
+
+func max(a, b uint8) uint8 {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+func (p *partialResult4ApproxCountDistinct) bufSize() uint32 {
+	return uint32(1) << p.sizeDegree
+}
+
+func (p *partialResult4ApproxCountDistinct) mask() uint32 {
+	return p.bufSize() - 1
+}
+
+func (p *partialResult4ApproxCountDistinct) place(x approxCountDistinctHashValue) uint32 {
+	return uint32(x>>uniquesHashBitsForSkip) & p.mask()
+}
+
+// Increase the size of the buffer 2 times or up to new size degree.
+func (p *partialResult4ApproxCountDistinct) resize(newSizeDegree uint8) {
+	oldSize := p.bufSize()
+	oldBuf := p.buf
+
+	if 0 == newSizeDegree {
+		newSizeDegree = p.sizeDegree + 1
+	}
+
+	p.buf = make([]approxCountDistinctHashValue, uint32(1)<<newSizeDegree)
+	p.sizeDegree = newSizeDegree
+
+	// Move some items to new locations.
+	for i := uint32(0); i < oldSize; i++ {
+		x := oldBuf[i]
+		if x != 0 {
+			p.reinsertImpl(x)
+		}
+	}
+}
+
+func (p *partialResult4ApproxCountDistinct) readAndMerge(rb []byte) error {
+	rhsSkipDegree := rb[0]
+	rb = rb[1:]
+
+	if rhsSkipDegree > p.skipDegree {
+		p.skipDegree = rhsSkipDegree
+		p.rehash()
+	}
+
+	rb, rhsSize, err := codec.DecodeUvarint(rb)
+
+	if err != nil {
+		return err
+	}
+
+	if rhsSize > uint64(uniquesHashMaxSize) {
```

Why this is checked here?

Should we return an error when rhsSize exceeding `uniquesHashMaxSize` and shrink can not be done?

---

This check is to avoid corner case when engine return illegal partial result.
In fact, it should not happen.